### PR TITLE
IBP-4207 Delete Germplasm

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

File: src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmServiceImpl.java [hunk 1]
```diff
@@ -371,21 +377,39 @@ public long countGermplasmMatches(final GermplasmMatchRequestDto germplasmMatchR
...
 		}
-		return germplasmService.findGermplasmMatches(germplasmMatchRequestDto, pageable);
+		return this.germplasmService.findGermplasmMatches(germplasmMatchRequestDto, pageable);
+	}
+
+	@Override
+	public GermplasmDeleteResponse deleteGermplasm(final List<Integer> gids) {
+
+		this.errors = new MapBindingResult(new HashMap<>(), String.class.getName());
+		this.germplasmValidator.validateGids(this.errors, gids);
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmServiceImpl.java:401]: If any of the GIDs are invalid or non-existing then the whole request will fail?

File: src/main/java/org/ibp/api/java/impl/middleware/keysequenceregister/KeySequenceRegisterServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.stream.Collectors;
+
+@Service
+@Transactional
+public class KeySequenceRegisterServiceImpl implements KeySequenceRegisterService {
+
+	private static final String SEQUENCE_NUMBER_REGEX = ")\\s?(\\d+).*";
+
+	@Resource
+	private GermplasmValidator germplasmValidator;
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/keysequenceregister/KeySequenceRegisterServiceImpl.java:27]: `germplasmValidator` seems to be unused

File: src/main/java/org/ibp/api/rest/keysequenceregister/KeySequenceRegisterResource.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+@Api(value = "Key Sequence Register Services")
+@Controller
+public class KeySequenceRegisterResource {
+
+	@Resource
+	private KeySequenceRegisterService keySequenceRegisterService;
+
+	@ApiOperation(value = "Delete Key Prefix Cache")
+	@PreAuthorize("hasAnyAuthority('ADMIN', 'CROP_MANAGEMENT', 'GERMPLASM', 'MANAGE_GERMPLASM')")
+	@RequestMapping(value = "/crops/{cropName}/keysequence", method = RequestMethod.DELETE)
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/keysequenceregister/KeySequenceRegisterResource.java:29]: resource name should be plural. how about `key-sequences`?

File: src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmServiceImpl.java (hunk 1 above)
Comment (abatac) [reply src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmServiceImpl.java:401]: If any of the GIDs are non-existing then the whole request will fail. But if all GIDs exist and some have validation error (has inventory, fixed group, etc..), the request will continue but only GIDs who passed the validation will be deleted.

File: src/main/java/org/ibp/api/java/keysequenceregister/KeySequenceRegisterService.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package org.ibp.api.java.keysequenceregister;
+
+import org.ibp.api.domain.keysequenceregister.KeySequenceRegisterDeleteResponse;
+
+import java.util.List;
+
+public interface KeySequenceRegisterService {
+
+	KeySequenceRegisterDeleteResponse deleteKeySequenceReqister(List<Integer> gids, List<String> prefixes);
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/keysequenceregister/KeySequenceRegisterService.java:9]: Better to rename as `deleteKeySequence`

File: src/main/java/org/ibp/api/rest/keysequenceregister/KeySequenceRegisterResource.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+public class KeySequenceRegisterResource {
+
+	@Resource
+	private KeySequenceRegisterService keySequenceRegisterService;
+
+	@ApiOperation(value = "Delete Key Prefix Cache")
+	@PreAuthorize("hasAnyAuthority('ADMIN', 'CROP_MANAGEMENT', 'GERMPLASM', 'MANAGE_GERMPLASM')")
+	@RequestMapping(value = "/crops/{cropName}/key-sequences", method = RequestMethod.DELETE)
+	@ResponseBody
+	public ResponseEntity<KeySequenceRegisterDeleteResponse> deleteGermplasm(@PathVariable final String cropName,
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/keysequenceregister/KeySequenceRegisterResource.java:31]: Change method name

File: src/main/java/org/ibp/api/rest/keysequenceregister/KeySequenceRegisterResource.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import javax.annotation.Resource;
+import java.util.List;
+
+@Api(value = "Key Sequence Register Services")
+@Controller
+public class KeySequenceRegisterResource {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/keysequenceregister/KeySequenceRegisterResource.java:22]: Rename as KeySequenceResource

File: src/main/java/org/ibp/api/rest/keysequenceregister/KeySequenceResource.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import javax.annotation.Resource;
+import java.util.List;
+
+@Api(value = "Key Sequence Register Services")
+@Controller
+public class KeySequenceResource {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/keysequenceregister/KeySequenceResource.java:22]: Logged https://ibplatform.atlassian.net/browse/IBP-4391 to create unit test for this class

**Review by darla-leafnode — APPROVED** [review]

**Pull request merged**